noir-projects/aztec-nr/aztec/src/macros/authorization.nr

```diff
@@ -0,0 +1,72 @@
+use crate::macros::utils::{
+    compute_struct_selector, derive_serialize_if_not_implemented, get_trait_impl_method,
+};
+use poseidon::poseidon2::Poseidon2Hasher;
+use std::{collections::umap::UHashMap, hash::BuildHasherDefault};
+
+pub comptime mut global AUTH_TYPES: UHashMap<Field, TypeDefinition, BuildHasherDefault<Poseidon2Hasher>> =
+    UHashMap::default();
+
+comptime fn register_authorization(
+    authorization_selector: Field,
+    authorization_struct: TypeDefinition,
+) {
+    if AUTH_TYPES.contains_key(authorization_selector) {
+        let existing_authorization = AUTH_TYPES.get(authorization_selector).unwrap().name();
+        let authorization_name = authorization_struct.name();
+        panic(
+            f"Selector collision detected between authorizations '{authorization_name}' and '{existing_authorization}'",
+        );
+    }
+    AUTH_TYPES.insert(authorization_selector, authorization_struct);
+}
+
+pub comptime fn generate_authorization_interface_and_get_selector(
+    s: TypeDefinition,
+) -> (Quoted, Field) {
+    let name = s.name();
+
+    let authorization_selector = compute_struct_selector(
+        s,
+        quote { crate::authwit::auth_selector::AuthorizationSelector::from_signature },
+    );
+
+    let from_field = get_trait_impl_method(
+        quote { crate::authwit::auth_selector::AuthorizationSelector }.as_type(),
+        quote { crate::protocol_types::traits::FromField },
+        quote { from_field },
+    );
+
+    let authorization_interface = quote { crate::authwit::auth_interface::AuthorizationInterface };
+    let authorization_selector_type =
+        quote { crate::authwit::auth_selector::AuthorizationSelector };
+
+    (
+        quote {
+        impl $authorization_interface for $name {
+            fn get_authorization_selector(self) -> $authorization_selector_type {
+                $from_field($authorization_selector)
+            }
+        }
+    },
+        authorization_selector,
+    )
+}
+
+/// An Authorization is a struct that represents and action a user can allow others to do
+/// on their behalf. By definition, an Authorization must be human-readable and contain
+/// enough information to be interpreted on its own. Authorizations are then hashed in a
+/// specific way to produce an inner_hash that can later be signed and checked via the
+/// authwit mechanism.
+pub comptime fn authorization(s: TypeDefinition) -> Quoted {
```

Comment:
This should explain in what contexts a dev would want to use this and how to use this with an example use case. I see that in the code it used only once on a struct called `CallAuthorization` which is not very helpful.

Reply:
Improved